iceoryx_posh/source/roudi/process_manager.cpp

```diff
@@ -0,0 +1,685 @@
+            // depending on the mode we clean up the process resources and register it again
+            // if it is monitored, we reject the registration and wait for automatic cleanup
+            // otherwise we remove the process ourselves and register it again
+
+            if (process.isMonitored()) // is it the same process or a duplicate?
+            {
+                // process exists and is monitored - we rely on monitoring for removal
+                LogWarn() << "Received REG from " << name
+                          << ", but another application with this name is already registered";
+
+                // Notify new application that it shall shutdown and try later
+                runtime::IpcMessage sendBuffer;
+                sendBuffer << runtime::IpcMessageTypeToString(
+                    runtime::IpcMessageType::REG_FAIL_RUNTIME_NAME_ALREADY_REGISTERED);
+                process.sendViaIpcChannel(sendBuffer);
+                returnValue = false;
+            }
+            else
+            {
+                // process exists and is not monitored - remove it and add the new process afterwards
+                LogDebug() << "Registering already existing application " << name;
+
+                // remove existing process
+                if (!searchForProcessAndRemoveIt(name))
+                {
+                    LogWarn()
+                        << "Received REG from " << name
+                        << ", but another application with this name is already registered and could not be removed";
+                    return;
+                }
+                else
+                {
+                    LogDebug() << "Removed existing application " << name;
+                    // try registration again, should succeed since removal was successful
+                    addProcess(name,
+                               pid,
+                               segmentInfo.m_memoryManager,
+                               isMonitored,
+                               transmissionTimestamp,
+                               segmentInfo.m_segmentID,
+                               sessionId,
+                               versionInfo);
+                    returnValue = true;
+                }
+            }
+        },
+        [&]() {
+            // process does not exist in list and can be added
+            addProcess(name,
+                       pid,
+                       segmentInfo.m_memoryManager,
+                       isMonitored,
+                       transmissionTimestamp,
+                       segmentInfo.m_segmentID,
+                       sessionId,
+                       versionInfo);
+            returnValue = true;
+        });
+
+    return returnValue;
+}
+
+bool ProcessManager::addProcess(const ProcessName_t& name,
+                                const uint32_t pid,
+                                cxx::not_null<mepoo::MemoryManager* const> payloadMemoryManager,
```

Comment:
please use a `mepoo::MemoryManager&` instead

Reply:
You're definitely right. However, this would result in moving the `nullptr` check one level up into `ProcessManager::registerProcess()`. The issue is that `SegmentUserInformation.m_memoryManager` is a pointer. To address that, it would need some bigger rework. I would suggest to stick with the nullptr check inside `addProcess` for now, otherwise it would be done all the time when calling `registerProcess` e.g. with `cxx::Expects(segmentInfo.m_memoryManager != nullptr)`

Reply:
We should probably use `std::reference_wrapper` in `SegmentUserInformation`. Could you either do that or create an issue for it/add a task somewhere?

It's used just a couple of times and would probably be quite easy to change.

Reply:
Add to #611